README
Win32-Monitoring-DllInject 
==========================

INSTALLATION

    Unpack the distribution into a local directory, and from a command
    prompt in that directory follow the instructions below.

    These instructions assume that you have already set up your build
    environment, and that you can build source distributions. The 
    following environments are known to work:

  Microsoft Visual Studio 2008

    Other versions of Visual Studio may work, adjust Makefile.PL 
    as necessary for your build environment

      perl Makefile.PL
      nmake
      nmake test
      nmake install

INJECTOR DLL 

    Our DLLInject.dll has known to compile with mingw32 (IMCROSS 2008-06-04)
    on Ubuntu 8.04 LTS compiled in the following way: 

       cd HookedFunctions
       make -f Makefile.migw32.mak

    Compiling with Visual Studio 2008 should work too after redefining some
    preprocessor instructions. This is still pending in the current
    HookedFunctions.c code.

SUPPORT

    Homepage: <http://oss.oetiker.ch/optools/>.

AUTHORS

    Roman Plessl,
    Tobias Oetiker

COPYRIGHT

    Copyright (c) 2008 by OETIKER+PARTNER AG. All rights reserved.

LICENSE

    Win32::Monitoring::DllInject is free software: you can redistribute 
    it and/or modify it under the terms of the GNU General Public License 
    as published by the Free Software Foundation, either version 3 of the 
    License, or (at your option) any later version.

    Win32::Monitoring::DllInject is distributed in the hope that it will 
    be useful, but WITHOUT ANY WARRANTY; without even the implied warranty 
    of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Win32::Monitoring::DllInject. If not, see 
    <http://www.gnu.org/licenses/>.

SEE ALSO

    Homepage: http://oss.oetiker.ch/optools/



Changes
Revision history for Perl extension Win32::Monitoring::DLLInject.

Win32-Monitoring-DllInject-0.05, Mon 2008-08-27
-----------------------------------------------
- Renamded DllInject to DLLInject
- Added README at HookedFunctions

Win32-Monitoring-DllInject-0.04, Mon 2008-08-25
-----------------------------------------------
- Improve monitoring example and code of the injecting DLL
- Fix correct handling of time stamps in DlInjector Example code
- moved test.pl to example.pl 
- added more documentation 

Win32-Monitoring-DllInject-0.03, Wed 2008-08-13 
-----------------------------------------------
- bugfixing of XS module and perl test program
- added license informations

Win32-Monitoring-DllInject-0.02, Thu 2008-07-22
----------------------------------------------- 
- XS code rewritten, now it does not use DllInject.h and DllInject.c anymore
- Changed Loading Interface of DllInject.pm
- Changed Makefile.PL

Win32-Monitoring-DllInject-0.01, Wed 2008-07-21
----------------------------------------------- 
- original version; created by h2xs 1.23 with options
  -Oan Win32::Monitoring::DllInject DllInject.h


HookedFunctions/test_llu.c
/* *********************************************************************
 * test_llu.c: Example for test the %llu behaviour of printf
 * *********************************************************************
 * Authors: Roman Plessl
 *
 * Copyright (c) 2008 by OETIKER+PARTNER AG. All rights reserved.
 * 
 * Win32::Monitoring::DllInject is free software: you can redistribute 
 * it and/or modify it under the terms of the GNU General Public License 
 * as published by the Free Software Foundation, either version 3 of the 
 * License, or (at your option) any later version.
 *
 * $Id: test_llu.c 143 2008-08-25 13:26:35Z rplessl $ 
 *
 * **********************************************************************
 */

#define __MSVCRT_VERSION__ 0x601
#define WINVER 0x0500 

#include <windows.h>
#include <stdio.h>
#include <tchar.h>
#include <stdlib.h>
#include <shlobj.h>
#include <imagehlp.h>

#include <sys/types.h>
#include <sys/timeb.h>
#include <time.h>

typedef unsigned long long   longtime_t;

longtime_t timems (void){
     struct __timeb64 timebuffer;
     _ftime64( &timebuffer );
     return ((longtime_t)timebuffer.time * 1000 + (longtime_t)timebuffer.millitm);
}

int main(void) {
  longtime_t now;
  now = timems(); 
  printf( "print unsigned long long I64u: %I64u\n", now ); 
  printf( "print unsigned long long llu: %llu\n", now );
}


example.pl
#! perl

use lib 'blib/lib';
use lib 'blib/arch';

use Win32::OLE;
use Win32::Monitoring::DllInject;
use Data::Dumper;

my $WshShell = Win32::OLE->new("WScript.Shell");
$WshShell->Run("notepad", 5);

sleep(1);

my %processes;

for my $line (`tasklist /v /nh`) {
 chomp($line);
 if ( $line ne "" ) {
  # extract PID
  my $pid = substr($line, 26, 8);
  # remove leading spaces
  $pid =~ s/^ *([0-9]+)$/$1/g;

  # extract process
  my $proc = substr($line, 0, 24);#.substr($line, 152, 72);
  # change multiple spaces to single spaces
  $proc =~ s/\s\s\s*/ /g;
  # remove trailing space
  $proc =~ s/\s$//g;
  # remove trailing N/A
  $proc =~ s/ N\/A$//g;

  # print tab seperated fields
  # print $pid, " ", $proc, "\n";
  $processes{$proc} = $pid;
  } 
}

my $P = Win32::Monitoring::DllInject->new($processes{'notepad.exe'},'Y:\\perl\\Win32-Monitoring-DllInject\\HookedFunctions.dll');

print Dumper($P);

while(1)
{
    sleep(1);
    my $msg_cnt = $P->StatMailSlot();
    for (my $i = 0; $i < $msg_cnt; $i++) {
        print $P->GetMessage(), "\n";
     }
}




Makefile.PL
use 5.008008;
use ExtUtils::MakeMaker;

use strict;
use warnings;
use Config;
#use Data::Dumper;
#map {$Config{$_} and $Config{$_} =~ /MD/ and print "$_ = $Config{$_}\n"} keys %Config;
my $CCFLAGS = $Config{ccflags};
my $OPTIMIZE = $Config{optimize};
$CCFLAGS =~ s/MD/MT/;
$OPTIMIZE =~ s/MD/MT/;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Win32::Monitoring::DllInject',
    VERSION_FROM      => 'DllInject.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'DllInject.pm', # retrieve abstract from module
       AUTHOR         => 'Roman Plessl (rplessl at cpan dot org)'
    ) : ()),
    INC               => '-I.', # e.g., '-I. -I/usr/include/other',
    CCFLAGS   => $CCFLAGS,
    OPTIMIZE => $OPTIMIZE,
    LIBS => ':nodefault :nosearch kernel32.lib user32.lib libcmt.lib',
    # Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);


DLLInject.pm
# $Id: DLLInject.pm 157 2008-08-27 08:06:41Z rplessl $

package Win32::Monitoring::DLLInject;

use 5.008008;
use strict;
use warnings;
use Carp;

require Exporter;
require DynaLoader;
use AutoLoader;

our @ISA = qw(Exporter DynaLoader);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Win32::Monitoring::DLLInject ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
   new
   StatMailSlot
   GetMessage
   Unhook
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw();

our $VERSION = '0.05';

bootstrap Win32::Monitoring::DLLInject $VERSION;

# Preloaded methods go here.

sub new {
    my $class = shift;
    my $self  =
        {
         'VERSION' => $VERSION,
        };

    bless $self, $class;

    $self->{pid} = shift;
    $self->{dll} = shift;

    ( $self->{hmailslot},
      $self->{hmodule},
      $self->{hooked} ) = init( $self->{pid}, $self->{dll} );

    return $self;
}

sub StatMailSlot {
    my $self = shift;
    return StatMailslot( $self->{hmailslot} );
}

sub GetMessage {
    my $self = shift;
    return GetMailslotMessage( $self->{hmailslot} );
}

sub Unhook {
   my $self = shift;

   ( $self->{hmodule},
     $self->{unhooked} ) = destroy($self->{pid}, $self->{dll}, $self->{hmailslot}, $self->{hmodule});

   if ($self->{unhooked} == 1) {
      $self->{hooked} = 0;
      delete $self->{unhooked};
   }
}

DESTROY {
   my $self = shift;
   ( $self->{hmodule},
     $self->{unhooked} ) = destroy($self->{pid}, $self->{dll}, $self->{hmailslot}, $self->{hmodule});

   if ($self->{unhooked} == 1) {
      $self->{hooked} = 0;
      delete $self->{unhooked};
   }
}


# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__

=head1 NAME

Win32::Monitoring::DLLInject - Injects Win32 programs with overloaded functions

=head1 SYNOPSIS

  use Win32::Monitoring::DLLInject qw(new UnHook StatMailslot GetMessage);

  my $handle = new Win32::Monitoring::DLLInject($dll_path, $process_id);

  while(1){
        sleep(1);
        my $msg_cnt = $handle->StatMailSlot();

        for (my $i = 0; $i < $msg_cnt; $i++) {
           print $handle->GetMessage(), "\n";
        }

  }
  $handle->UnHook();

=head1 DESCRIPTION

The Win32::Monitoring::DLLInject module provides a perl object to automatically
handles and injects a Windows program or a DLL with some overloading (self written)
functional code.

Additional a communication infrastructure is set up using a Windows mailslot
to return information like status information or time measurements to the callee.

As a bonus there is an example framework for a DLL implementation included, such
that you can implement a nice time measuring monitoring program for any Win32
application without requiring further modules.

=over

=item $handle = new($dll_path,$process_id)

Returns an handle to the Win32::Monitoring::DLLInject object to handle the
overloaded (hooked) program.

=item $handle->StatMailSlot()

Returns the amount of messages in the internal message store (mailslot).

=item $handle->GetMessage()

Returns the content of the first message in the message store.

=item $handle->Unhook()

Reverts the hooking of the program injecting.

=back

=head2 EXAMPLE

  #! perl

  use Win32::OLE;
  use Win32::Monitoring::DLLInject;
  use Data::Dumper;

  my $WshShell = Win32::OLE->new("WScript.Shell");
  $WshShell->Run("notepad", 5);

  sleep(1);

  my %processes;

  for my $line (`tasklist /v /nh`) {
     chomp($line);
     if ( $line ne "" ) {
        my $pid = substr($line, 26, 8);  # extract PID
        $pid =~ s/^ *([0-9]+)$/$1/g;     # remove leading spaces

        my $proc = substr($line, 0, 24); # extract process
        $proc =~ s/\s\s\s*/ /g;          # change multiple spaces to single spaces
        $proc =~ s/\s$//g;               # remove trailing space
        $proc =~ s/ N\/A$//g;            # remove trailing N/A

        $processes{$proc} = $pid;
      }
  }

  my $P = Win32::Monitoring::DLLInject->new($processes{'notepad.exe'},'Y:\\perl\\Win32-Monitoring-DLLInject\\HookedFunctions.dll');

  print Dumper($P);

  while(1)
  {
       sleep(1);
       my $msg_cnt = $P->StatMailSlot();
       for (my $i = 0; $i < $msg_cnt; $i++) {
           print $P->GetMessage(), "\n";
       }
  }

=head1 SEE ALSO

Webpage: <http://oss.oetiker.ch/optools/>

=head1 COPYRIGHT

Copyright (c) 2008 by OETIKER+PARTNER AG. All rights reserved.

=head1 LICENSE

Win32::Monitoring::DLLInject is free software: you can redistribute
it and/or modify it under the terms of the GNU General Public License
as published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

Win32::Monitoring::DLLInject is distributed in the hope that it will
be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Win32::Monitoring::WindowPing. If not, see
<http://www.gnu.org/licenses/>.


=head1 AUTHORS

Roman Plessl,
Tobi Oetiker

=cut


HookedFunctions/HookedFunctions.c
/* *********************************************************************
 * Win32::Monitoring::DllInject - 
 *    Injects Win32 programs with overloaded functions
 * *********************************************************************
 * HookedFunctions.c: Template and example for injecting DLL
 * *********************************************************************
 * Based on code by Parag Paithankar  
 * http://www.codeproject.com/KB/system/api_monitoring_unleashed.aspx 
 * *********************************************************************
 * Authors: Tobias Oetiker
 *          Roman Plessl
 *
 * Copyright (c) 2008 by OETIKER+PARTNER AG. All rights reserved.
 * 
 * Win32::Monitoring::DllInject is free software: you can redistribute 
 * it and/or modify it under the terms of the GNU General Public License 
 * as published by the Free Software Foundation, either version 3 of the 
 * License, or (at your option) any later version.
 *
 * $Id: WindowPing.xs 127 2008-08-13 09:09:45Z rplessl $ 
 *
 * **********************************************************************
 */

#define __MSVCRT_VERSION__ 0x601
#define WINVER 0x0500 

#include <windows.h>
#include <stdio.h>
#include <tchar.h>
#include <stdlib.h>
#include <shlobj.h>
#include <imagehlp.h>

/* include this for time measurements */
#define TIMESTAMP

/* include this for debugging with message boxes */
//#define VERBOSE


/* *** MAILSLOT FOR COMMUNICATION *** */
#define __MAILSLOT__ _TEXT("\\\\.\\mailslot\\F9AD51D8-428A-11DD-9E14-7E7256D89593")

/* remarks to the define myllu:

   myllu is a 'fix' for the 'not correctly' processing of unsigned long long
   in msvcr.dll. this should be added/fixed in msvcr80.dll, but I have not
   tested this. The linux printf llu is I64u according to:

     http://msdn.microsoft.com/en-us/library/tcxf1dw6.aspx

   You can test the behaviour of your system by building and running the
   test_llu.exe

     make test_llu.exe
     test_llu.exe
   
*/
/* *** Redefines for Visual C Compiler *** */
#ifdef _MSC_VER

#define myLPCWSTR  LPCSTR
#define myLLU      L"%I64u"

#else 

#define myLPCWSTR  LPCWSTR
#define myLLU      L"%I64u"

#endif

/* *** GLOBAL VARIABLES *** */
myLPCWSTR SlotName = __MAILSLOT__;
DWORD CurrentProcessID;  

/* *** TIMESTAMP HANDLING *** */

#ifdef TIMESTAMP
#include <sys/types.h>
#include <sys/timeb.h>
#include <time.h>

typedef unsigned long long   longtime_t;

longtime_t timems (void){
     struct __timeb64 timebuffer;
     _ftime64( &timebuffer );
     return ((longtime_t)timebuffer.time * 1000 + (longtime_t)timebuffer.millitm);
}

#endif


/* *** FUNCTIONS *** */
BOOL WriteSlot( HANDLE hSlot, LPCWSTR lpszMessage )
{
   static HANDLE s_hSlot = INVALID_HANDLE_VALUE;
   /* call once with hSlot set to the real handle
      to initialize WriteSlot */   

   if (hSlot != INVALID_HANDLE_VALUE){
       s_hSlot = hSlot;
   }

   if (s_hSlot != INVALID_HANDLE_VALUE){
       DWORD cbWritten; 
#ifdef VERBOSE
#ifdef _MSC_VER
       MessageBoxW(NULL, lpszMessage, _TEXT(L"Message"), MB_OK | MB_ICONINFORMATION);
#else
       MessageBox(NULL, lpszMessage, _TEXT("Message"), MB_OK | MB_ICONINFORMATION);
#endif
#endif
       return WriteFile(s_hSlot, 
                        lpszMessage, 
                        (DWORD) (wcslen(lpszMessage)+1)*sizeof(WCHAR),
                        &cbWritten, 
                        (LPOVERLAPPED) NULL); 
   }
   return 0;
}

void SetHook( HMODULE hModuleOfCaller, HMODULE hModule, LPSTR LibraryName, PROC OldFunctionPointer, PROC NewFunctionPointer )
{
    ULONG ulSize;
    PIMAGE_IMPORT_DESCRIPTOR pImportDesc = NULL;
    PSTR pszModName = NULL;
    PIMAGE_THUNK_DATA pThunk = NULL;
    PROC pfnCurrent = NULL;
    PROC* ppfn = NULL;
    BOOL bFound = FALSE;
    DWORD dwOldProtect = 0;

    // track process id of current process cause the same library
    // can be used to inject multiple programs
    CurrentProcessID = GetCurrentProcessId();
    
    if(hModuleOfCaller == hModule)
        return;
    if(hModuleOfCaller == 0)
        return;

    // Get the address of the module's import section
    pImportDesc = (PIMAGE_IMPORT_DESCRIPTOR) ImageDirectoryEntryToData
    (
        hModuleOfCaller, 
        TRUE, 
        IMAGE_DIRECTORY_ENTRY_IMPORT, 
        &ulSize
    );

    // Does this module have an import section ?
    if (pImportDesc == NULL)
        return;

    // Loop through all descriptors and find the 
    // import descriptor containing references to callee's functions
    while (pImportDesc->Name)
    {
        pszModName = (PSTR)((PBYTE) hModuleOfCaller + pImportDesc->Name);
        
        if (strcmp(pszModName, LibraryName) == 0) 
            break; // Found

        pImportDesc++;
    } // while

    if (pImportDesc->Name == 0)
        return;

    //Get caller's IAT 
    pThunk = (PIMAGE_THUNK_DATA)( (PBYTE) hModuleOfCaller + pImportDesc->FirstThunk );

    pfnCurrent = OldFunctionPointer;

    // Replace current function address with new one
    while (pThunk->u1.Function)
    {
        // Get the address of the function address
        ppfn = (PROC*) &pThunk->u1.Function;
        // Is this the function we're looking for?
        bFound = (*ppfn == pfnCurrent);

        if (bFound) 
        {
            MEMORY_BASIC_INFORMATION mbi;
            
            VirtualQuery(ppfn, &mbi, sizeof(MEMORY_BASIC_INFORMATION));

            // In order to provide writable access to this part of the 
            // memory we need to change the memory protection

            if (FALSE == VirtualProtect(mbi.BaseAddress,mbi.RegionSize,PAGE_READWRITE,&mbi.Protect))
                return;

            *ppfn = NULL;
            *ppfn = *NewFunctionPointer;

            // Restore the protection back
            dwOldProtect = 0;

            VirtualProtect(mbi.BaseAddress,mbi.RegionSize,mbi.Protect,&dwOldProtect);
            
            break;
        } // if

        pThunk++;

    } // while
}

PROC EnumAndSetHooks( LPSTR BaseLibraryName, LPSTR BaseFunctionName, PROC NewFunctionPointer, HMODULE hModule, BOOL UnHook, PROC Custom )
{
    HMODULE hMods[1024];
    DWORD cbNeeded;
    unsigned int i;
    typedef BOOL (WINAPI * PFNENUMPROCESSMODULES)
    (
        HANDLE hProcess,
        HMODULE *lphModule,
        DWORD cb,
        LPDWORD lpcbNeeded
    );

    PROC hBaseProc;

    HMODULE hBaseLib = NULL;
    PFNENUMPROCESSMODULES m_pfnEnumProcessModules = NULL;
    HMODULE m_hModPSAPI = NULL;
    HANDLE hProcess = NULL;
   
    hBaseLib = LoadLibraryA(BaseLibraryName);

    if(UnHook)
        hBaseProc = (PROC) Custom;
    else
        hBaseProc = GetProcAddress(hBaseLib, BaseFunctionName);

    m_hModPSAPI = LoadLibraryA("PSAPI.DLL");

    m_pfnEnumProcessModules = (PFNENUMPROCESSMODULES)GetProcAddress(m_hModPSAPI, "EnumProcessModules");

    hProcess = GetCurrentProcess();

    if( m_pfnEnumProcessModules(hProcess, hMods, sizeof(hMods), &cbNeeded)) {
        for ( i = 0; i < (cbNeeded / sizeof(HMODULE)); i++ ) {
            SetHook(hMods[i], hModule, BaseLibraryName, hBaseProc, NewFunctionPointer); 
        }
    }

    return hBaseProc;
}

/* wrapper for return value functions */
#define fnwrap(return_type, function, arg_def, arg_call, arg_monitor) \
        static PROC g_Original ## function = NULL; \
        typedef return_type WINAPI w32apimon_ ## function ## _t arg_def; \
        return_type WINAPI w32apimon_ ## function arg_def { \
            return_type ReturnValue; \
            unsigned long long start; unsigned long long now; \
            WCHAR starttime[24]; WCHAR deltatime[24]; WCHAR processid[8]; WCHAR functionname[50]; WCHAR measurement[300]; \
            start = timems(); \
            w32apimon_ ## function ## _t* fn = (w32apimon_ ## function ## _t*)g_Original  ## function; \
            ReturnValue = (*fn) arg_call; \
            now = timems(); \
            swprintf( starttime,    myLLU,                 start ); \
            swprintf( deltatime,    myLLU,                 (now-start) ); \
            swprintf( processid,    L"%u",                 CurrentProcessID ); \
            swprintf( functionname, L"%S",                 #function ); \
            swprintf( measurement,  L"%s\t%s\t%s\t%s\t%s", starttime, deltatime, processid, functionname, arg_monitor); \
            WriteSlot(INVALID_HANDLE_VALUE,measurement); \
            return ReturnValue; \
        }

/* wrapper for void functions */
#define vfnwrap(function, arg_def, arg_call) \
        static PROC g_Original ## function = NULL; \
        typedef void WINAPI w32apimon_ ## function ## _t arg_def; \
        void WINAPI w32apimon_ ## function arg_def {\
            longtime_t start; longtime_t now; \
            WCHAR starttime[24]; WCHAR deltatime[24]; WCHAR processid[8]; WCHAR functionname[50];  WCHAR measurement[300]; \
            start = timems(); \
            w32apimon_ ## function ## _t* fn = (w32apimon_ ## function ## _t*)g_Original  ## function; \
            (*fn) arg_call; \
            now = timems(); \
            swprintf( starttime,    myLLU,                 start ); \
            swprintf( deltatime,    myLLU,                 (now-start) ); \
            swprintf( processid,    L"%u",                 CurrentProcessID ); \
            swprintf( functionname, L"%S",                 #function ); \
            swprintf( measurement,  L"%s\t%s\t%s\t%s\t%s", starttime, deltatime, processid, functionname); \
            WriteSlot(INVALID_HANDLE_VALUE,measurement);                     

#define hook(dll,function) \
            g_Original ## function = EnumAndSetHooks(#dll, #function, (PROC) w32apimon_ ## function, hModule, FALSE, 0);

#define unhook(dll,function) \
            EnumAndSetHooks(#dll, #function, (PROC) ( g_Original ## function != NULL ? g_Original ## function : GetProcAddress(LoadLibraryA(#dll),#function)), hModule, TRUE, (PROC) w32apimon_ ## function);

/***************************************************************/
/* Here are the declaration for the the replacment functions   */
/* BEGIN replacement functions prototypes                      */
/***************************************************************/
fnwrap(BOOL,SetCurrentDirectoryW,(LPCWSTR lpPathName),(lpPathName),lpPathName)

fnwrap(BOOL,CreateDirectoryW,(LPCWSTR lpPathName, LPSECURITY_ATTRIBUTES lpSecurityAttributes),(lpPathName,lpSecurityAttributes),lpPathName)

vfnwrap(SHAddToRecentDocs,( UINT uFlags, LPCVOID pv),(uFlags,pv))
            if (uFlags & SHARD_PATH){
                    WriteSlot(INVALID_HANDLE_VALUE,pv);
            } else {
                    WCHAR pszPath[MAX_PATH];
                    if (SHGetPathFromIDListW(pv,pszPath)){
                        WriteSlot(INVALID_HANDLE_VALUE,pszPath);
                    }
            }
}            
/***************************************************************/
/* END replacement functions prototypes                        */
/***************************************************************/

BOOL APIENTRY DllMain( HANDLE hModule, DWORD ul_reason_for_call, LPVOID lpReserved)
{
    HANDLE hSlot = NULL;
    switch (ul_reason_for_call)
    {
        case DLL_PROCESS_ATTACH:
            hSlot = CreateFile(SlotName, 
                               GENERIC_WRITE, 
                               FILE_SHARE_READ,
                               (LPSECURITY_ATTRIBUTES) NULL, 
                               OPEN_EXISTING, 
                               FILE_ATTRIBUTE_NORMAL, 
                               (HANDLE) NULL); 

/***************************************************************/
/* replacement for these functions in the corresponding DLL    */
/* BEGIN function replacement (hook)                           */
/***************************************************************/

            hook(KERNEL32.dll,SetCurrentDirectoryW)
            hook(KERNEL32.dll,CreateDirectoryW)
            hook(shell32.dll,SHAddToRecentDocs)

/***************************************************************/
/* END function replacement (hook)                             */
/***************************************************************/

#ifdef _MSC_VER
            WriteSlot(hSlot,TEXT(L"Functions Hooked"));            
#else
            WriteSlot(hSlot,TEXT("Functions Hooked"));  
#endif

            break;
        case DLL_PROCESS_DETACH:

/***************************************************************/
/* replacement for these functions in the corresponding DLL    */
/* BEGIN function replacement (unhook)                         */
/***************************************************************/           

            unhook(KERNEL32.dll,SetCurrentDirectoryW)
            unhook(KERNEL32.dll,CreateDirectoryW)
            unhook(shell32.dll,SHAddToRecentDocs)

/***************************************************************/
/* END function replacement (unhook)                           */
/***************************************************************/

#ifdef _MSC_VER
            WriteSlot(hSlot,TEXT(L"Functions UnHooked"));            
#else
            WriteSlot(hSlot,TEXT("Functions UnHooked"));
#endif

            break;
    }
    return TRUE;
}



DllInject.pm
# $Id: DllInject.pm 143 2008-08-25 13:26:35Z rplessl $

package Win32::Monitoring::DllInject;

use 5.008008;
use strict;
use warnings;
use Carp;

require Exporter;
require DynaLoader;
use AutoLoader;

our @ISA = qw(Exporter DynaLoader);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Win32::Monitoring::DllInject ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
   new
   StatMailSlot
   GetMessage
   Unhook
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw();

our $VERSION = '0.04';

bootstrap Win32::Monitoring::DllInject $VERSION;

# Preloaded methods go here.

sub new {
    my $class = shift;
    my $self  =
        {
         'VERSION' => $VERSION,
        };

    bless $self, $class;

    $self->{pid} = shift;
    $self->{dll} = shift;

    ( $self->{hmailslot},
      $self->{hmodule},
      $self->{hooked} ) = init( $self->{pid}, $self->{dll} );

    return $self;
}

sub StatMailSlot {
    my $self = shift;
    return StatMailslot( $self->{hmailslot} );
}

sub GetMessage {
    my $self = shift;
    return GetMailslotMessage( $self->{hmailslot} );
}

sub Unhook {
   my $self = shift;

   ( $self->{hmodule},
     $self->{unhooked} ) = destroy($self->{pid}, $self->{dll}, $self->{hmailslot}, $self->{hmodule});

   if ($self->{unhooked} == 1) {
      $self->{hooked} = 0;
      delete $self->{unhooked};
   }
}

DESTROY {
   my $self = shift;
   ( $self->{hmodule},
     $self->{unhooked} ) = destroy($self->{pid}, $self->{dll}, $self->{hmailslot}, $self->{hmodule});

   if ($self->{unhooked} == 1) {
      $self->{hooked} = 0;
      delete $self->{unhooked};
   }
}


# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__

=head1 NAME

Win32::Monitoring::DllInject - Injects Win32 programs with overloaded functions

=head1 SYNOPSIS

  use Win32::Monitoring::DllInject qw(new UnHook StatMailslot GetMessage);

  my $handle = new Win32::Monitoring::DllInject($dll_path, $process_id);

  while(1){
        sleep(1);
        my $msg_cnt = $handle->StatMailSlot();

        for (my $i = 0; $i < $msg_cnt; $i++) {
           print $handle->GetMessage(), "\n";
        }

  }
  $handle->UnHook();

=head1 DESCRIPTION

The Win32::Monitoring::DllInject module provides a perl object to automatically
handles and injects a Windows program or a DLL with some overloading (self written)
functional code.

Additional a communication infrastructure is set up using a Windows mailslot
to return information like status information or time measurements to the callee.

As a bonus there is an example framework for a DLL implementation included, such
that you can implement a nice time measuring monitoring program for any Win32
application without requiring further modules.

=over

=item $handle = new($dll_path,$process_id)

Returns an handle to the Win32::Monitoring::DllInject object to handle the
overloaded (hooked) program.

=item $handle->StatMailSlot()

Returns the amount of messages in the internal message store (mailslot).

=item $handle->GetMessage()

Returns the content of the first message in the message store.

=item $handle->Unhook()

Reverts the hooking of the program injecting.

=back

=head2 EXAMPLE

  #! perl

  use Win32::OLE;
  use Win32::Monitoring::DllInject;
  use Data::Dumper;

  my $WshShell = Win32::OLE->new("WScript.Shell");
  $WshShell->Run("notepad", 5);

  sleep(1);

  my %processes;

  for my $line (`tasklist /v /nh`) {
     chomp($line);
     if ( $line ne "" ) {
        my $pid = substr($line, 26, 8);  # extract PID
        $pid =~ s/^ *([0-9]+)$/$1/g;     # remove leading spaces

        my $proc = substr($line, 0, 24); # extract process
        $proc =~ s/\s\s\s*/ /g;          # change multiple spaces to single spaces
        $proc =~ s/\s$//g;               # remove trailing space
        $proc =~ s/ N\/A$//g;            # remove trailing N/A

        $processes{$proc} = $pid;
      }
  }

  my $P = Win32::Monitoring::DllInject->new($processes{'notepad.exe'},'Y:\\perl\\Win32-Monitoring-DllInject\\HookedFunctions.dll');

  print Dumper($P);

  while(1)
  {
       sleep(1);
       my $msg_cnt = $P->StatMailSlot();
       for (my $i = 0; $i < $msg_cnt; $i++) {
           print $P->GetMessage(), "\n";
       }
  }

=head1 SEE ALSO

Webpage: <http://oss.oetiker.ch/optools/>

=head1 COPYRIGHT

Copyright (c) 2008 by OETIKER+PARTNER AG. All rights reserved.

=head1 LICENSE

Win32::Monitoring::DllInject is free software: you can redistribute
it and/or modify it under the terms of the GNU General Public License
as published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

Win32::Monitoring::DllInject is distributed in the hope that it will
be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Win32::Monitoring::WindowPing. If not, see
<http://www.gnu.org/licenses/>.


=head1 AUTHORS

Roman Plessl,
Tobi Oetiker

=cut
